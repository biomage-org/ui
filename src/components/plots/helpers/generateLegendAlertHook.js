import _ from 'lodash';

const MAX_LEGEND_ITEMS = 50;

const generateLegendAlertHook = (hierarchy, cellSetKey, useConfigValue = true) => {
  const hookFn = (plotConfig) => {
    const cellSetName = useConfigValue ? plotConfig[cellSetKey] : cellSetKey;

    const numLegendItems = hierarchy.find(
      ({ key }) => key === cellSetName,
    )?.children?.length;

    if (numLegendItems < MAX_LEGEND_ITEMS) return plotConfig;

    const modifiedLegend = {
      enabled: false,
      showAlert: true,
    };

    if (!plotConfig.legend.enabled) modifiedLegend.showAlert = false;

    const modifiedConfig = _.merge({}, plotConfig, {
      legend: modifiedLegend,
    });

    return modifiedConfig;
  };

  return hookFn;
};

export default generateLegendAlertHook;
export { MAX_LEGEND_ITEMS };
