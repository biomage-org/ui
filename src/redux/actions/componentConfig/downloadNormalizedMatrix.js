// import { PLOT_DATA_LOADED, PLOT_DATA_LOADING, PLOT_DATA_ERROR }
// from 'redux/actionTypes/componentConfig';

import { fetchWork } from 'utils/work/fetchWork';
// import handleError from 'utils/http/handleError';
// import endUserMessages from 'utils/endUserMessages';

const downloadNormalizedMatrix = (
  experimentId,
  subsetBy,
) => async (dispatch, getState) => {
  // try {
  const body = {
    name: 'GetNormalizedExpression',
    subsetBy,
  };

  const data = await fetchWork(
    experimentId, body, getState,
  );

  console.log('dataDebug');
  console.log(data);
  // } catch (e) {
  //   const errorMessage = handleError(e, endUserMessages.ERROR_FETCHING_PLOT_DATA);

  //   // dispatch({
  //   //   type: PLOT_DATA_ERROR,
  //   //   payload: {
  //   //     plotUuid,
  //   //     error: errorMessage,
  //   //   },
  //   // });
  // }
};

export default downloadNormalizedMatrix;
