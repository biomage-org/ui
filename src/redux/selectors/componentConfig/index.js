import getPlotConfigs from './getPlotConfigs';
import getSelectedMetadataTracks from './getSelectedMetadataTracks';

/* eslint-disable-next-line import/prefer-default-export */
export { getPlotConfigs, getSelectedMetadataTracks };
