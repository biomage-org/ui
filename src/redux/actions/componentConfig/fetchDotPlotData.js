import { initialPlotConfigStates } from 'redux/reducers/componentConfig/initialState';
import getTimeoutForWorkerTask from 'utils/getTimeoutForWorkerTask';
import { PLOT_DATA_LOADED, PLOT_DATA_LOADING, PLOT_DATA_ERROR } from 'redux/actionTypes/componentConfig';

import handleError from 'utils/http/handleError';
import endUserMessages from 'utils/endUserMessages';
import generatePlotWorkBody from 'utils/work/generatePlotWorkBody';
import { fetchWork } from 'utils/work/fetchWork';

const getClusterNames = (state) => {
  const clusterIds = state.cellSets.hierarchy.reduce(
    (acc, cellSet) => [...acc, ...cellSet.children.map((entry) => entry.key)],
    [],
  );

  const clusterNames = clusterIds?.map((clusterId) => state.cellSets.properties[clusterId].name);
  return clusterNames;
};

const orderCellSets = (data, state, config) => {
  // reordering data based on the sample order
  const { selectedCellSet } = config;
  const { hierarchy, properties } = state.cellSets;
  if (hierarchy.length) {
    const cellSetOrderKeys = hierarchy.filter((rootNode) => rootNode.key === selectedCellSet)[0]
      .children
      .map((cellSet) => cellSet.key);
    const cellSetOrderNames = cellSetOrderKeys.map((cellSet) => properties[cellSet].name);
    data.sort((a, b) => (
      cellSetOrderNames.indexOf(a.cellSets) - cellSetOrderNames.indexOf(b.cellSets)
    ));
  }
};

const fetchDotPlotData = (
  experimentId,
  plotUuid,
  plotType,
) => async (dispatch, getState) => {
  let config = getState().componentConfig[plotUuid]?.config ?? initialPlotConfigStates[plotType];
  const clusterNames = getClusterNames(getState());
  const timeout = getTimeoutForWorkerTask(getState(), 'PlotData');

  config = {
    ...config,
    clusterNames,
  };

  try {
    const body = generatePlotWorkBody(plotType, config);

    dispatch({
      type: PLOT_DATA_LOADING,
      payload: { plotUuid },
    });

    const data = await fetchWork(
      experimentId, body, getState, { timeout },
    );

    orderCellSets(data, getState(), config);

    dispatch({
      type: PLOT_DATA_LOADED,
      payload: {
        plotUuid,
        plotData: data,
      },
    });
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_FETCHING_PLOT_DATA);

    dispatch({
      type: PLOT_DATA_ERROR,
      payload: {
        plotUuid,
        error: errorMessage,
      },
    });
  }
};

export default fetchDotPlotData;
