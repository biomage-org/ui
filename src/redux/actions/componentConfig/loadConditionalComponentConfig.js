import _ from 'lodash';
import fetchAPI from 'utils/http/fetchAPI';
import endUserMessages from 'utils/endUserMessages';
import { LOAD_CONFIG } from 'redux/actionTypes/componentConfig';
import { initialComponentConfigStates, initialPlotConfigStates } from 'redux/reducers/componentConfig/initialState';
import handleError from 'utils/http/handleError';
import httpStatusCodes from 'utils/http/httpStatusCodes';

const loadConditionalComponentConfig = (
  experimentId, componentUuid, type, skipAPI, customConfig = {},
) => async (dispatch) => {
  const defaultConfig = initialPlotConfigStates[type] ?? initialComponentConfigStates[type];
  const configToUse = _.merge({}, defaultConfig, customConfig);

  const dispatchLoad = (config) => {
    dispatch({
      type: LOAD_CONFIG,
      payload: {
        experimentId,
        plotUuid: componentUuid,
        plotType: type,
        plotData: [],
        config,
      },
    });
  };

  if (skipAPI) {
    dispatchLoad(configToUse);
    return;
  }

  try {
    const data = await fetchAPI(`/v2/experiments/${experimentId}/plots/${componentUuid}`);

    const plotConfig = _.merge({}, configToUse, data.config);
    dispatchLoad(plotConfig);
  } catch (e) {
    // load default plot config if it not found
    if (e.statusCode === httpStatusCodes.NOT_FOUND) {
      dispatchLoad(configToUse);
      return;
    }

    handleError(e, endUserMessages.ERROR_FETCHING_PLOT_CONFIG);
  }
};

export default loadConditionalComponentConfig;
