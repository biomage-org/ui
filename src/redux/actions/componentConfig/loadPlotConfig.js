import _ from 'lodash';
import fetchAPI from 'utils/http/fetchAPI';
import endUserMessages from 'utils/endUserMessages';
import { LOAD_CONFIG } from 'redux/actionTypes/componentConfig';
import { initialPlotConfigStates } from 'redux/reducers/componentConfig/initialState';
import handleError from 'utils/http/handleError';
import httpStatusCodes from 'utils/http/httpStatusCodes';

const loadPlotConfig = (
  experimentId,
  plotUuid,
  plotType,
  beforeLoadConfigHook,
) => async (dispatch) => {
  try {
    const { config, plotData } = await fetchAPI(`/v2/experiments/${experimentId}/plots/${plotUuid}`);

    let plotConfig = _.merge({}, initialPlotConfigStates[plotType], config);
    plotConfig = beforeLoadConfigHook ? beforeLoadConfigHook(plotConfig) : plotConfig;

    dispatch({
      type: LOAD_CONFIG,
      payload: {
        experimentId,
        plotUuid,
        plotType,
        plotData,
        config: plotConfig,
      },
    });
  } catch (e) {
    // load default plot config if it not found
    if (e.statusCode === httpStatusCodes.NOT_FOUND) {
      let plotConfig = _.cloneDeep(initialPlotConfigStates[plotType]);
      if (beforeLoadConfigHook) plotConfig = beforeLoadConfigHook(plotConfig);

      dispatch({
        type: LOAD_CONFIG,
        payload: {
          experimentId,
          plotUuid,
          plotType,
          plotData: [],
          config: plotConfig,
        },
      });
      return;
    }

    handleError(e, endUserMessages.ERROR_FETCHING_PLOT_CONFIG);
  }
};

export default loadPlotConfig;
