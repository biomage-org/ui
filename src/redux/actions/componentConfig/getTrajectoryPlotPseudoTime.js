import getTimeoutForWorkerTask from 'utils/getTimeoutForWorkerTask';
import { PLOT_DATA_LOADED, PLOT_DATA_LOADING, PLOT_DATA_ERROR } from 'redux/actionTypes/componentConfig';

import handleError from 'utils/http/handleError';
import endUserMessages from 'utils/endUserMessages';
import { fetchWork, generateETag } from 'utils/work/fetchWork';
import { getBackendStatus } from 'redux/selectors';

const getPseudoTime = (
  rootNodes,
  experimentId,
  plotUuid,
) => async (dispatch, getState) => {
  const {
    embeddingSettings,
    clusteringSettings,
  } = getState().experimentSettings.processing?.configureEmbedding || {};

  const embeddingState = getState()
    .experimentSettings
    ?.processing
    ?.configureEmbedding
    ?.embeddingSettings;

  const {
    methodSettings,
    method: embeddingMethod,
  } = embeddingState;

  const { environment } = getState().networkResources;
  const backendStatus = getBackendStatus(experimentId)(getState()).status;
  const { pipeline: { startDate: qcPipelineStartDate } } = backendStatus;

  const embeddingBody = {
    name: 'GetEmbedding',
    type: embeddingMethod,
    config: methodSettings[embeddingMethod],
  };

  const timeout = getTimeoutForWorkerTask(getState(), 'TrajectoryAnalysisPseudotime');

  const embeddingETag = generateETag(
    experimentId,
    embeddingBody,
    undefined,
    qcPipelineStartDate,
    environment,
  );

  const body = {
    name: 'GetPseudoTime',
    embedding: {
      method: embeddingSettings.method,
      ETag: embeddingETag,
    },
    clustering: {
      method: clusteringSettings.method,
      resolution: clusteringSettings.methodSettings[clusteringSettings.method].resolution,
    },
    rootNodes,
  };

  try {
    dispatch({
      type: PLOT_DATA_LOADING,
      payload: { plotUuid },
    });

    const data = await fetchWork(
      experimentId, body, getState, { timeout, rerun: true },
    );

    const { plotData } = getState().componentConfig[plotUuid];

    dispatch({
      type: PLOT_DATA_LOADED,
      payload: {
        plotUuid,
        plotData: {
          ...plotData,
          pseudotime: data.pseudotime,
        },
      },
    });
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_FETCHING_PLOT_DATA);

    dispatch({
      type: PLOT_DATA_ERROR,
      payload: {
        plotUuid,
        error: errorMessage,
      },
    });
  }
};

export default getPseudoTime;
