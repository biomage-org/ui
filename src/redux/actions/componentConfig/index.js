import loadPlotConfig from 'redux/actions/componentConfig/loadPlotConfig';
import loadComponentConfig from 'redux/actions/componentConfig/loadComponentConfig';
import updatePlotConfig from 'redux/actions/componentConfig/updatePlotConfig';
import savePlotConfig from 'redux/actions/componentConfig/savePlotConfig';
import fetchDotPlotData from 'redux/actions/componentConfig/fetchDotPlotData';
import updatePlotData from 'redux/actions/componentConfig/updatePlotData';
import resetPlotConfig from 'redux/actions/componentConfig/resetPlotConfig';

export {
  loadPlotConfig,
  loadComponentConfig,
  updatePlotConfig,
  savePlotConfig,
  fetchDotPlotData,
  updatePlotData,
  resetPlotConfig,
};
