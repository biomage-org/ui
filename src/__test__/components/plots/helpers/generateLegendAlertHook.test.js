import generateLegendAlertHook, { MAX_LEGEND_ITEMS } from 'components/plots/helpers/generateLegendAlertHook';
import { legendBaseState } from 'redux/reducers/componentConfig/baseStylesState';

const hierarchy = [
  {
    key: 'louvain',
    children: [...Array(100)].map((_, i) => `louvain-${i}`),
  },
  {
    key: 'sample',
    children: [...Array(100)].map((_, i) => `sample-${i}`),
  },
];

const config = {
  selectedCellSet: 'louvain',
  legend: legendBaseState,
};

describe('generateLegendAlertHook', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Should return a function', () => {
    const hookFn = generateLegendAlertHook(hierarchy, 'selectedCellSet');

    expect(typeof hookFn).toEqual('function');
  });

  it('Should use the plotConfig cellSet value for comparison', () => {
    const hookFn = generateLegendAlertHook(hierarchy, 'selectedCellSet');

    const modifiedConfig = hookFn(config);

    expect(JSON.stringify(config)).not.toEqual(JSON.stringify(modifiedConfig));
    expect(modifiedConfig).toMatchSnapshot();
  });

  it('Should be able to use the hardcoded value for comparison', () => {
    const hookFn = generateLegendAlertHook(hierarchy, 'sample', false);

    const modifiedConfig = hookFn(config);

    expect(JSON.stringify(config)).not.toEqual(JSON.stringify(modifiedConfig));
    expect(modifiedConfig).toMatchSnapshot();
  });

  it('Should not show legendAlert if plotConfig is already hidden', () => {
    const hiddenConfig = {
      selectedCellSet: 'louvain',
      legend: {
        enabled: false,
        showAlert: true,
      },
    };

    const hookFn = generateLegendAlertHook(hierarchy, 'selectedCellSet');

    const modifiedConfig = hookFn(hiddenConfig);

    expect(modifiedConfig.legend.showAlert).toEqual(false);
  });

  it('Should return original config if number of cellSets is less than maximum to be shown', () => {
    const smallHierarchy = [{
      key: 'louvain',
      children: [...Array(10)].map((_, i) => `louvain-${i}`),
    }];

    const hookFn = generateLegendAlertHook(smallHierarchy, 'selectedCellSet');

    const modifiedConfig = hookFn(config);

    expect(JSON.stringify(config)).toEqual(JSON.stringify(modifiedConfig));
  });

  it('Exports the number of max legend items', () => {
    expect(MAX_LEGEND_ITEMS).toMatchInlineSnapshot('50');
  });
});
