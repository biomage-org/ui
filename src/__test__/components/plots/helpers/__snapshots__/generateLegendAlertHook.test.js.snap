// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateLegendAlertHook Should be able to use the hardcoded value for comparison 1`] = `
Object {
  "legend": Object {
    "colour": "#000000",
    "enabled": false,
    "position": "right",
    "showAlert": true,
  },
  "selectedCellSet": "louvain",
}
`;

exports[`generateLegendAlertHook Should use the plotConfig cellSet value for comparison 1`] = `
Object {
  "legend": Object {
    "colour": "#000000",
    "enabled": false,
    "position": "right",
    "showAlert": true,
  },
  "selectedCellSet": "louvain",
}
`;
